eclipse/src/saros/editor/internal/ContributionAnnotationManager.java

```diff
@@ -168,33 +172,54 @@ public void splitAnnotation(final IAnnotationModel model, final int offset) {
 
       if ((offset > pos.offset) && (offset < pos.offset + pos.length)) {
 
-        Position before = new Position(pos.offset, offset - pos.offset);
+        final Position before = new Position(pos.offset, offset - pos.offset);
 
-        Position after = new Position(offset, pos.length - (offset - pos.offset));
+        final Position after = new Position(offset, pos.length - (offset - pos.offset));
 
         annotationsToRemove.add(contributionAnnotation);
 
-        annotationsToAdd.put(new ContributionAnnotation(source, model), before);
+        final ContributionAnnotation beforeAnnotation = new ContributionAnnotation(source, model);
+        final ContributionAnnotation afterAnnotation = new ContributionAnnotation(source, model);
 
-        annotationsToAdd.put(new ContributionAnnotation(source, model), after);
-      }
-    }
+        annotationsToAdd.put(beforeAnnotation, before);
+        annotationsToAdd.put(afterAnnotation, after);
+
+        final History history = getHistory(source);
+
+        for (final ListIterator<HistoryEntry> lit = history.entries.listIterator();
+            lit.hasNext(); ) {
```

Comment:
This could be a for-each loop instead of manually advancing the iterator.

Reply:
Please look carefully at the code again, especially the code starting at line 203. This is not a simple iteration.

Reply:
Okay right, then please add some comments to the code.
It took me way too long to grasp, what this is doing (especially because I have no idea about annotations) and I think this could be explained. (I also feel like this could be way easier using functional-programming, but that's my personal opinion.)

Reply:
Functional Programming is normally harder to read.

removeAllAnnotation ended up with:

    sourceToHistory
        .values()
        .stream()
        .flatMap(h -> h.entries.stream())
        .collect(
            HashSet<IAnnotationModel>::new,
            (s, e) -> s.add(e.annotation.getModel()),
            HashSet::addAll)
        .forEach(
            m ->
                annotationModelHelper.removeAnnotationsFromModel(
                    m, (a) -> a instanceof ContributionAnnotation));


On the other hand optimization is done by manipulating the existing data structure.  That is something that you do not do in functional programming, e.g you have a big list where some entries should be removed. In a functional approach you create a new list containing all the entries without the entries that should be removed. While this may be fine for an ArrayList in terms of performance it is a huge performance loss for a LinkedList.

Reply:
> Functional Programming is normally harder to read.

I would in general not agree with that statement. One could even argue your suggested functional code is more readable, because it does the same, while being shorter.

But I get, that a stream of streams is generally not what you want. (Which is why I just requested some comments on your old approach.)

> While this may be fine for an ArrayList in terms of performance it is a huge performance loss for a LinkedList.

While this is obviously true, this, in my opinion, an optimization the compiler should be able to infer.

Reply:
The Java compiler does not really optimize anything. Even the JIT does not optimize much stuff. You are simply expecting too much from a compiler.

Reply:
Not from a compiler, just from Java/JVM...

Reply:
Read again. Just In Time (compiler) ... it is part of the JVM.

Really the can not optimize every thing.

for (int i = 0; i < x; i++)
{
   Collection c = xyz.calculate(...) // will run in O(n^10) and always return the same result
   ... do something with c
}

I have seen such cases where the JIT is unable to move such calls out of the loop because the compiler fails to determine if calculate is idempotent in the current context and so you end up with a total running time of O(x * n^10) instead of O(x + n^10).